src/spec.rs

```diff
@@ -153,56 +500,105 @@ pub struct Zip64CentralDirectoryEnd {
+            disk_number,
+            disk_with_central_directory,
+            number_of_files_on_this_disk,
+            number_of_files,
+            central_directory_size,
+            central_directory_offset,
+        })
+    }
+
     pub fn find_and_parse<T: Read + Seek>(
         reader: &mut T,
         nominal_offset: u64,
         search_upper_bound: u64,
     ) -> ZipResult<Vec<(Zip64CentralDirectoryEnd, u64)>> {
         let mut results = Vec::new();
-        let mut pos = search_upper_bound;
-
-        while pos >= nominal_offset {
-            let mut have_signature = false;
-            reader.seek(io::SeekFrom::Start(pos))?;
-            if reader.read_u32_le()? == ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE {
-                have_signature = true;
-                let archive_offset = pos - nominal_offset;
-
-                let _record_size = reader.read_u64_le()?;
-                // We would use this value if we did anything with the "zip64 extensible data sector".
-
-                let version_made_by = reader.read_u16_le()?;
-                let version_needed_to_extract = reader.read_u16_le()?;
-                let disk_number = reader.read_u32_le()?;
-                let disk_with_central_directory = reader.read_u32_le()?;
-                let number_of_files_on_this_disk = reader.read_u64_le()?;
-                let number_of_files = reader.read_u64_le()?;
-                let central_directory_size = reader.read_u64_le()?;
-                let central_directory_offset = reader.read_u64_le()?;
-
-                results.push((
-                    Zip64CentralDirectoryEnd {
-                        version_made_by,
-                        version_needed_to_extract,
-                        disk_number,
-                        disk_with_central_directory,
-                        number_of_files_on_this_disk,
-                        number_of_files,
-                        central_directory_size,
-                        central_directory_offset,
-                    },
-                    archive_offset,
-                ));
+
+        const END_WINDOW_SIZE: usize = 2048;
+        /* TODO: use static_assertions!() */
+        assert!(END_WINDOW_SIZE > mem::size_of::<Magic>());
+
+        const SIG_BYTES: [u8; mem::size_of::<Magic>()] =
+            Magic::ZIP64_CENTRAL_DIRECTORY_END_SIGNATURE.to_le_bytes();
+        let finder = FinderRev::new(&SIG_BYTES);
+
+        let mut window_start: u64 = search_upper_bound
+            .saturating_sub(END_WINDOW_SIZE as u64)
+            .max(nominal_offset);
+        let mut window = [0u8; END_WINDOW_SIZE];
+        while window_start >= nominal_offset {
+            reader.seek(io::SeekFrom::Start(window_start))?;
+
+            /* Identify how many bytes to read (this may be less than the window size for files
+             * smaller than END_WINDOW_SIZE). */
+            let end = (window_start + END_WINDOW_SIZE as u64).min(search_upper_bound);
```

Comment:
> Is this correct even for the old definition of `search_upper_bound` (see my comment on line 308)? If not, take out the min.

The comment on line 308 refers to limiting the value of `search_lower_bound` when it is not already provided. In this method, we have been provided a `search_lower_bound`, so the comment on line 308 does not apply. I have renamed variables appropriately to make this more clear.

Reply:
> Doesn't this potentially make the last window unnecessarily long when it comes at the beginning of the file? 

Yes, but since we are relying on `memchr` to perform fast SIMD searching and our window size is only 512 bytes, it seems very unlikely that this is going to become a performance concern. The current calculation mirrors the work done for the zip32 case, and I think maintaining that consistency is likely to be more readable and easier to maintain than attempting to prematurely optimize here.